text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,82 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC introduces a new concept called the "constness-context" of an execution of a function. A `const fn` is executed in a const constness-context if it was called in a context where only `const fn` are allowed (like `static` initializers) or inside another `const fn` that was executed in a const constness-context. All other calls to a `const fn` (for example in `main`) are in a runtime constness-context.
+
+We can therefore say that code can either be:
+- Always in a const constness-context (for example `static` initializers, array lengths...)
+- Maybe in a const constness-context (`const fn`), where the constness-context can differ between calls depending on the call-site
+- Always in a runtime const constness-context (non-`const` functions)
+
+A `const fn` is now allowed to exhibit different behavior depending on it being called in a const or runtime constness-context. Language features and standard library functions may also differ in behavior depending on the constness-context they have been used or called in.
+
+This makes the constness-context of a function observable behavior.
+
+If a `const fn` is called in a runtime constness-context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+A `const fn` being called in a const constness-context will still be required to be deterministic, as this is required for type systme soundness.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime constness-context, as long as it behaves exactly like runtime. The optimizer could also manually annotate functions as being truly pure by looking at the body.
+
+Secondly, with the current rules around `const fn` purity, unsafe code could choose to only sometimes call unknown `const fn` code at runtime depending on whether it needs the result. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time". This could break code that already relies on this, but since Rust doesn't have proper support for this, the impact should be minimal.
```

Comment:
```suggestion
Secondly, with the current rules around `const fn` purity, unsafe code could choose rely on purity, e.g. by caching function return values and assuming this is not observable to clients. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time". This could break code that already relies on this, but since Rust doesn't have proper support for this, the impact should be minimal.
```